pulsar-client-go/examples/consumer/consumer.go

```diff
@@ -0,0 +1,54 @@
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+package main
+
+import (
+	"../../pulsar"
+	"fmt"
+	"log"
+)
+
+func main() {
+	client := pulsar.NewClient().
+		ServiceUrl("pulsar://localhost:6650").Build()
+
+	consumer, err := client.NewConsumer().
+		Topic("my-topic").
+		SubscriptionName("my-subscription").
+		SubscriptionType(pulsar.Shared).
+		Subscribe()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for {
+		msg, err := consumer.Receive()
```

Comment:
Is there a per-message timeout by chance? If so, then `Receive` could benefit from passing `context.Context` which is another idiom for libraries doing I/O where you may want or need to specific a deadline for the work.

Reply:
`Receive()` would block (on a channel) until a message is available. there's also `ReceiveWithTimeout()`. Is that what you are referring with " per-message timeout " ?

Reply:
Yes. So rather than having `Receive` and `ReceiveWithTimeout`, you could have `Receive(context.Context)` and that context may have a deadline set already and handle both use cases. In the implementation you would check this:

```go
// Use timeout if deadline is set..
if d, ok := ctx.Deadline(); ok {
  timeout := d.Sub(time.Now())
  // ..
}

// Otherwise, block indefinitely.
```

Reply:
Actually, I've come to conclusion to remove `Receive()` and `Next()` with timeouts. If one wants to have that, the channel listener can be used.